controller/aks-cluster-config-handler.go

```diff
@@ -0,0 +1,728 @@
+	}
+
+	baseURL := spec.BaseURL
+	if baseURL == "" {
+		baseURL = azure.PublicCloud.ResourceManagerEndpoint
+	}
+
+	client := resources.NewGroupsClientWithBaseURI(baseURL, spec.SubscriptionID)
+	client.Authorizer = authorizer
+
+	return &client, nil
+}
+
+func createResourceGroup(ctx context.Context, groupsClient *resources.GroupsClient, spec aksv1.AKSClusterConfigSpec) (bool, error) {
+	future, err := groupsClient.CreateOrUpdate(
+		ctx,
+		spec.ResourceGroup,
+		resources.Group{
+			Name:	  to.StringPtr(spec.ResourceGroup),
+			Location: to.StringPtr(spec.ResourceLocation),
+		})
+
+	if err != nil {
+		return false, fmt.Errorf("[AKS] error creating resource group '%s': %v", spec.ResourceGroup, err)
+	}
+
+	return future.StatusCode == 201, nil
+}
+
+func checkResourceGroup(ctx context.Context, groupsClient *resources.GroupsClient, spec aksv1.AKSClusterConfigSpec) (bool, error) {
+	resp, err := groupsClient.CheckExistence(ctx, spec.ResourceGroup)
+	if err != nil {
+		return false, fmt.Errorf("[AKS] error getting resource group '%s': %v", spec.ResourceGroup, err)
+	}
+
+	return resp.StatusCode == 204, nil
+}
+
+func removeResourceGroup(ctx context.Context, groupsClient *resources.GroupsClient, spec aksv1.AKSClusterConfigSpec) error {
+	future, err := groupsClient.Delete(ctx, spec.ResourceGroup)
+	if err != nil {
+		return fmt.Errorf("[AKS] error removing resource group '%s': %v", spec.ResourceGroup, err)
+	}
+
+	if err = future.WaitForCompletionRef(ctx, groupsClient.Client); err != nil {
+		logrus.Errorf("[AKS] cannot get the AKS cluster create or update future response: %v", err)
+		return err
+	}
+
+	logrus.Infof("Cluster %v removed successfully", spec.ClusterName)
+
+	return nil
+}
+
+func newClustersClient(authorizer autorest.Authorizer, spec aksv1.AKSClusterConfigSpec) (*containerservice.ManagedClustersClient, error) {
+	if authorizer == nil {
+		return nil, fmt.Errorf("error getting Azure authorizer")
+	}
+
+	baseURL := spec.BaseURL
+	if baseURL == "" {
+		baseURL = azure.PublicCloud.ResourceManagerEndpoint
+	}
+
+	client := containerservice.NewManagedClustersClientWithBaseURI(baseURL, spec.SubscriptionID)
+	client.Authorizer = authorizer
+
+	return &client, nil
+}
+
+// CreateAKS creates a new managed Kubernetes cluster
+func createOrUpdateCluster(ctx context.Context, secretsCache wranglerv1.SecretCache, clusterClient *containerservice.ManagedClustersClient, spec aksv1.AKSClusterConfigSpec) (c containerservice.ManagedCluster, err error) {
+
+	dnsPrefix := spec.DNSPrefix
+	if dnsPrefix == "" {
+		dnsPrefix = spec.ClusterName + "-master"
+	}
+
+	tags := make(map[string]*string)
+	for key, val := range spec.Tags {
+		if val != "" {
+			tags[key] = to.StringPtr(val)
+		}
+	}
+	displayName := spec.DisplayName
+	if displayName == "" {
+		displayName = spec.ClusterName
+	}
+	tags["displayName"] = to.StringPtr(displayName)
+
+	kubernetesVersion := spec.KubernetesVersion
+	if kubernetesVersion == "" {
+		kubernetesVersion = "1.18.14"
+	}
```

Comment:
As in a previous comment with node pool, would the spec's k8s version ever be empty? Shouldn't that be based on the selection the user makes from a drop-down in the ui? The other issue I have here is AKS keeps removing support for specific patch versions so this might fail but this would be moot if it's based on a ui selection.

Reply:
+1

`KubernetesVersion` should not be blank, and setting a hard-coded default would result in constant updating of the default.